### PR TITLE
Add Gutenberg Embed WebView Activity

### LINKED ISSUE #5509: Video player
Add a native supported VideoPress video player block

### PR DESCRIPTION
**Pull request by jhnstn**
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5509

Adds a new embed WebView for the editor.

Related PRs:
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5743
- https://github.com/WordPress/gutenberg/pull/50440
- https://github.com/Automattic/jetpack/pull/30521

To test:
1. Apply both related PRs.
2. Open a post with a VideoPress block in the Editor
3. Select the VideoPress block and press again to play the preview video
4. The VideoPress embed player should open in a WebView
5. Exit the WebView via swipe or by pressing the back arrow. 
6. Expect to return to the editor.

## Regression Notes
1. Potential unintended areas of impact
Gutenberg bridge code - minimal.

7. What I did to test those areas of impact (or what existing automated tests I relied on)
Existing code was not altered for these changes. 

8. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [x] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: libs/editor/src/main/java/org/wordpress/android/editor/WPGutenbergEmbedWebViewActivity.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+public class WPGutenbergEmbedWebViewActivity extends GutenbergEmbedWebViewActivity {
+    @Override
+    protected void load() {
+        String content = getIntent().getExtras().getString(ARG_CONTENT);
+        mWebView.loadData(content, "text/html", "UTF-8");
+    }
+
+    protected String getToolbarTitle() {
+        return getIntent().getExtras().getString(ARG_TITLE);
+    }
```
Comment (fluiddot) [inline libs/editor/src/main/java/org/wordpress/android/editor/WPGutenbergEmbedWebViewActivity.java]: I see that these functions don't have specific logic that depends on WP-Android code like the Unsupported Block Editor WebView. Hence, probably we can do this in Gutenberg ([reference](https://github.com/WordPress/gutenberg/blob/19cdfa9f77b6720d0bf8ae316335e31022aacccf/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergEmbedWebViewActivity.java#L81-L105)) and we can avoid creating this activity here.

File: libs/editor/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -7,6 +7,9 @@
         <activity
             android:name=".WPGutenbergWebViewActivity"
             android:theme="@style/Base.GutenbergWebView" />
+        <activity
+            android:name=".WPGutenbergEmbedWebViewActivity"
+            android:theme="@style/Base.GutenbergWebView" />
```
Comment (fluiddot) [inline libs/editor/src/main/AndroidManifest.xml]: Following the above comment, if we remove `WPGutenbergEmbedWebViewActivity` class we can also remove this change.

**Review by fluiddot — COMMENTED** [review]
@jhnstn I tested the changes and work as expected. I added a couple of comments that shouldn't block the PR, let me know if you could take a look.

File: libs/editor/src/main/java/org/wordpress/android/editor/WPGutenbergEmbedWebViewActivity.java (hunk 1 above)
Comment (jhnstn) [reply libs/editor/src/main/java/org/wordpress/android/editor/WPGutenbergEmbedWebViewActivity.java]: Yeah I noticed how thin this extended  class got compared to the unsupported block webview. I wasn't sure how necessary it was to have this run on the Android side of things. 

I'll take this out.

Comment (fluiddot) [reply libs/editor/src/main/java/org/wordpress/android/editor/WPGutenbergEmbedWebViewActivity.java]: I tried this approach locally and seemed to work, so I think it's safe to remove it.

Comment (jhnstn) [reply libs/editor/src/main/java/org/wordpress/android/editor/WPGutenbergEmbedWebViewActivity.java]: fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/18383/commits/9ffa9f018cded839a61609af2c2803fe6f3e321a

File: libs/editor/src/main/AndroidManifest.xml (hunk 2 above)
Comment (jhnstn) [reply libs/editor/src/main/AndroidManifest.xml]: fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/18383/commits/9ffa9f018cded839a61609af2c2803fe6f3e321a

**Comment by fluiddot** [comment]
@jhnstn since this PR also enables the VideoPress block on Android, what do you think about updating the PR's title and description to reflect this?

**UPDATE:** Alternatively, we could enable the block in a separate PR.

**Review by fluiddot — COMMENTED** [review]
I tested the last installable build and noticed the following issues:

### R-rated videos
The keyboard remains open when introducing the DOB and closing the WebView. We'd need to figure out how to force closing the keyboard when going back from the WebView.

**NOTE:** I saw that we have a [similar issue when interacting with the prompt dialog](https://github.com/WordPress/gutenberg/issues/50105).

https://github.com/wordpress-mobile/WordPress-Android/assets/14905380/ec7d8120-6b1a-49d1-adf8-1ee75da75184

Additionally, I noticed that R-rated videos display the DOB prompt in the editor. Since the player will be only available via the WebView, I'm wondering if instead of rendering the preview content, we could simply display the poster image with a play button. I wouldn't consider this a blocker but more like an enhancement.

### Private videos
I checked that private videos are not displayed when opening a WebView, this issue is a bit critical as the player is not available. Hence, I'd consider it a blocker for releasing the block on Android.

In theory, the VideoPress URL set in the iframe's source should contain the token ([reference](https://github.com/Automattic/jetpack/blob/77ad8203862db43b7cc23ad6180ebe424849af1b/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.js#LL90C8-L90C12)), so not sure why private videos can't be played 🤔.

https://github.com/wordpress-mobile/WordPress-Android/assets/14905380/13672b05-1438-4c92-bd3f-4b74b0788ee5

**Comment by jhnstn** [comment]
Thanks for finding these @fluiddot  

Can you give any more info on the private video case? I've testing with private videos but can't reproduce:

https://github.com/wordpress-mobile/WordPress-Android/assets/744755/1e58dfa3-79de-4c96-b8f5-b33a38ea219a

That's on a simple site where videos are not private by default but that particular video is set as private. I've also tried with the site set to private. The player loads for me in both cases.

Regarding the mature settings, I expect being able to force close the keyboard should be possible when the WebView is closed. My feeling is that is not a blocker to this PR but also borderline as a blocker to releasing VP on Android.

**Comment by fluiddot** [comment]
> Can you give any more info on the private video case? I've testing with private videos but can't reproduce:

@jhnstn Yep, sorry I didn't share all the info. I reproduced it using a Premium simple site with site privacy set to `Private`. I can also reproduce it by setting the site `Public`. I tested using the [installable build `pr18383-144b55d`](https://github.com/wordpress-mobile/WordPress-Android/pull/18383#issuecomment-1539322571). In the video you shared, I noticed that it's in development mode by the RN error/warning messages. Can you try to reproduce it using the installable build?

_Tested with Samsung Galaxy S20 FE 5G (Android 13)._

> Regarding the mature settings, I expect being able to force close the keyboard should be possible when the WebView is closed. My feeling is that is not a blocker to this PR but also borderline as a blocker to releasing VP on Android.

I agree, I wouldn't consider it a blocker for the PR but a nice-to-have before releasing on Android.

**Review by fluiddot — APPROVED** [review]
> @jhnstn Yep, sorry I didn't share all the info. I reproduced it using a Premium simple site with site privacy set to `Private`. I can also reproduce it by setting the site `Public`. I tested using the [installable build `pr18383-144b55d`](https://github.com/wordpress-mobile/WordPress-Android/pull/18383#issuecomment-1539322571). In the video you shared, I noticed that it's in development mode by the RN error/warning messages. Can you try to reproduce it using the installable build?

I'm testing again and, after uploading a video and marking it as private, now I can't reproduce the issue 🙃. However, for the video I shared in the video captured above, I'm still not able to play it. For some reason, I'm having weird connection issues in the Android device I'm using, so I'm suspecting that might be related to that. That said, now that I can't reproduce it I'll approve the PR.

LGTM 🎊 !

**Comment by lmischner** [comment]
Talkback doesn't seem to work very well with the web view.
With Talkback on, when I open the web view, focus is on the close button (which seems fine), but, after reading the video file name, talkback says "play video, share button, doubletap activate'. This isn't accurate -- double tapping closes the video.
If I tap on the video itself, Talkback doesn't read anything so a low sighted user wouldn't know there's a video to play or how to do so.

**Comment by jhnstn** [comment]
@lmischner I'm not sure we can do much about these talkback issues. The only think we control in the WebView is the iframe. The player itself loads from VideoPress. 
We can use JavaScript to listen for play/pause events from the player but I'm not sure how we can feed those into the Android talkback system. I'm looking into that now.

**Pull request merged**